### PR TITLE
adding ability to pass in github token to dockerfile

### PR DESCRIPTION
**Pull request by dnomadb**
- Adds ability to pass in `GithubAccessToken` as a build-arg to a docker file per #76 

cc @jacquestardie @rclark

### CODE REVIEW

File: utils.sh [hunk 1]
```diff
@@ -81,6 +81,12 @@ function credentials() {
     args+="--build-arg NPMAccessToken=${NPMAccessToken}"
   fi
 
+  GithubAccessToken=$(printenv | grep GithubAccessToken | sed 's/.*=//')
+  if [[ -n $GithubAccessToken ]] && grep "ARG GithubAccessToken" ${filepath} > /dev/null 2>&1; then
+    args+="--build-arg GithubAccessToken=${GithubAccessToken}"
```
Comment (rclark) [inline utils.sh]: Watch out for spaces between `--build-arg`s

File: test/utils.test.sh [hunk 2]
```diff
@@ -238,19 +239,34 @@ clear_dockerfile
 credentials ${tmpdocker}
 assert "doesNotContain" "${args}" "NPMAccessToken=${NPMAccessToken}"
 
+# credentials() no github token in env test
```
Comment (rclark) [inline test/utils.test.sh]: GithubToken differs from NPMAccessToken in that the github token is a required env var -- you can't run conex without it. I think you should be able to remove this test, and simplify the condition under which you set the --build-arg in utils.sh

**Comment by dnomadb** [comment]
I haven't been able to get this to work; the main thing I am after is the ability to do:
```
ARG GithubAccessToken
...
RUN git clone https://$GithubAccessToken@github.com/<user>/<repo>.git
```
Within a `Dockerfile`. Here is me experimenting: https://github.com/mapbox/offset-detect/pull/1

**Comment by dnomadb** [comment]
This is failing on staging, with the error:
```
[Thu, 27 Oct 2016 23:38:51 GMT] [worker] [8099bf39-1556-4040-8e6e-756fa34ca630] checking environment configuration
[Thu, 27 Oct 2016 23:38:51 GMT] [worker] [8099bf39-1556-4040-8e6e-756fa34ca630] checking docker configuration
[Thu, 27 Oct 2016 23:38:51 GMT] [worker] [8099bf39-1556-4040-8e6e-756fa34ca630] checking job receive count
[Thu, 27 Oct 2016 23:38:51 GMT] [worker] [8099bf39-1556-4040-8e6e-756fa34ca630] parsing received message
[Thu, 27 Oct 2016 23:38:51 GMT] [worker] [8099bf39-1556-4040-8e6e-756fa34ca630] processing commit 30df266fea431bd226ca55c8ad35e145eb6e3470 by dnomadb to refs/heads/start-me-up of mapbox/offset-detect
[Thu, 27 Oct 2016 23:38:51 GMT] [worker] [8099bf39-1556-4040-8e6e-756fa34ca630] sending pending status to github
[Thu, 27 Oct 2016 23:38:52 GMT] [worker] [8099bf39-1556-4040-8e6e-756fa34ca630] fatal: could not read Password for 'https://github.com': No such device or address
```
I thought that this was from attempted cloning **in** the cloned dockerfile, but if you look at the above logs, it isn't even getting to that point; it is failing here: https://github.com/mapbox/ecs-conex/blob/0649508046cc2beeddf3b9229bccf33d74483a29/ecs-conex.sh#L35

This same repo is also being watched in production, and it passes this point, which leads me to believe that there might just be a problem with the staging stack's GithubAccessToken.

**Comment by rclark** [comment]
Been trying to get to the bottom of this and I've walked away quite confused. Clearly something is failing to decrypt the encrypted GithubAccessToken and NPMAccessToken on the staging stack, but the logs are quite confusing:

The watcher gets going...

```
> watchbot@1.0.3 start /usr/local/src/watchbot
> bin/watchbot.js
```

The watcher receives a job...
```
[Thu, 10 Nov 2016 18:45:30 GMT] [watchbot] [1b0984d3-825a-4bc4-b27d-3cd95e85c040] started task for webhook: {"ref":"refs/heads/several","after":"6bc98a152e2589722e94f480360b4f6dc1fd5689","before":"3f1f7add8c708d0472ae6e751443057dea3caf36","deleted":false,"repository":{"name":"ecs-telephone","owner":{"name":"mapbox"}},"pusher":{"name":"rclark"}}
```

A container starts up, and [the first thing it should do is decrypt environment variables](https://github.com/mapbox/ecs-conex/blob/c4c75feca25e41c1134b990b648e434d49f5e79c/Dockerfile#L31). Two weird things here are a) it should also decrypt the GithubAccessToken, and b) the lack of suffix on the `************` indicates that [the decrypted value was actually blank](https://github.com/mapbox/decrypt-kms-env/blob/9f1bdf3db00889957e572c1c4f4cf0f665f3a3c5/bin/decrypt-kms-env#L32).
```
Decrypted NPMAccessToken=************
```

Then we see worker logs from `ecs-conex.sh` kick in...
```
[Thu, 10 Nov 2016 18:46:06 GMT] [worker] [1b0984d3-825a-4bc4-b27d-3cd95e85c040] checking docker configuration
```

At this point I put in some debug logs to `printenv` and notice that:
- NPMAccessToken is blank
- GithubAccessToken is blank
- AWS_DEFAULT_REGION is not defined, which means the [decrypt-kms-env call should've failed hard](https://github.com/mapbox/decrypt-kms-env/blob/9f1bdf3db00889957e572c1c4f4cf0f665f3a3c5/bin/decrypt-kms-env#L10)

The worker continues to work, and a debug log indicates that the clone is not using any token...
```
[Thu, 10 Nov 2016 18:46:06 GMT] [worker] [1b0984d3-825a-4bc4-b27d-3cd95e85c040] checking environment configuration
[Thu, 10 Nov 2016 18:46:06 GMT] [worker] [1b0984d3-825a-4bc4-b27d-3cd95e85c040] sending pending status to github
[Thu, 10 Nov 2016 18:46:06 GMT] [worker] [1b0984d3-825a-4bc4-b27d-3cd95e85c040] git clone https://@github.com/mapbox/ecs-telephone /mnt/data/tS2Moh
```

Work continues because this particular repo is public. Conex logs into ECR and starts building an image...
```
[Thu, 10 Nov 2016 18:46:06 GMT] [worker] [1b0984d3-825a-4bc4-b27d-3cd95e85c040] looking for dockerfile
[Thu, 10 Nov 2016 18:46:06 GMT] [worker] [1b0984d3-825a-4bc4-b27d-3cd95e85c040] gather local credentials and setup --build-arg
[Thu, 10 Nov 2016 18:46:06 GMT] [worker] [1b0984d3-825a-4bc4-b27d-3cd95e85c040] logging into ECR repositories in us-east-1 us-west-2 eu-west-1
...
[Thu, 10 Nov 2016 18:46:10 GMT] [worker] [1b0984d3-825a-4bc4-b27d-3cd95e85c040] building new image
```

Then for some reason there are a couple of awscli configuration warnings wrapped around the missing GithubAccessToken decryption, which also failed (no suffix)
```
You must specify a region. You can also configure your region by running "aws configure".
Decrypted GithubAccessToken=************
You must specify a region. You can also configure your region by running "aws configure".
```

There's clearly something going on with decrypt-kms-env here that I don't understand yet. Will keep on digging.

**Comment by rclark** [comment]
Post #81, please try rebasing this over master and testing on staging again.

**Comment by perrygeo** [comment]
🆗  Pushed [this](https://github.com/mapbox/offset-detect/pull/1/files#diff-3254677a7917c6c01f55212f86c57fbfR23) through staging and confirmed that the github access token can be used as a build arg. This is good to go @rclark @dnomadb, anything else before we merge and tag 0.3?

**Comment by rclark** [comment]
LGTM

**Pull request merged**